XSConsoleLang.py

```diff
@@ -13,10 +13,36 @@
 # with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
+import sys
+
+import XenAPI  # For XenAPI.Failure
+
 from XSConsoleConfig import *
 from XSConsoleLangErrors import *
 
-import XenAPI # For XenAPI.Failure
+# Assign the unicode text type to a variable for use with isinstance():
+if sys.version_info >= (3, 0):
+    text_type = str
+else:
+    text_type = unicode  # pyright:ignore[reportUndefinedVariable] # pylint: disable=unicode-builtin
+
+
+def convert_anything_to_str(arg):
+    """Converts anything into the native "str" type of the Python version, without u'str' or b'str'"""
+    if arg is None or isinstance(arg, str):
+        return arg  # Already str or None (checked by some callers), return it as-is:
+
+    # If "unicode" text or "bytes", en- or decode accordingly:
+    if isinstance(arg, (text_type, bytes)):
+        if sys.version_info > (3, 0):
+            # Python3: Decode UTF-8 bytes into the native Python3 Unicode string:
+            return arg.decode("utf-8")
+        else:
+            # Python2: Encode the unicode text into the stream of UTF-8 bytes:
+            return arg.encode("utf-8")
```

Comment:
Are you sure the direction is correct here? I don't know much about unicode and python2 vs 3 but it seems wrong that one version you have to encode and on the other one you decode. You either end up double encoding something or double decoding something.

This would be a lot clearer if there were 2 if conditions, one for each type.
`bytes` -> `str` I assume this will be decode on both python 2 and 3
`unicode` -> `str` I assume this will be an encode

Reply:
No, it's the reverse (yes):

For reference: In the early 2000's I was the SuSE maintainer for the `pine` email Program, including its text edtior `pico`, and based on initial patches by i18n expert, I developed the SuSE patches to make it use UTF-8 as its native data format, and handle all conversion from and to all text encoding from E-Mail to and from UTF-8 as well as to and from the locale encoding of the terminal the user uses as well. So I'm quite familar with the topic.

The two data types at play here are:
- UTF-8 (a sequence of bytes): This is a string of single bytes, e.g.
  - UTF-8 is an encoding, the data is encoded "in" UTF-8 according to UTF-8 code tables
  - To convert from UTF-8 to something else, the UTF-8 bytes is `decoded` according to tables
- Unicode data type (called "str" in Python3)
  - This is a data type which hides the internals of each character from the user, so he does not have to deal with how the code points are actually stored. The user just sees "characters", which makes life dealing with non-ASCII strings a lot easier than dealing with strings encoded in various locale encodings, especially UTF-8, which encodes characters as variable-length sequences of bytes.

`.decode("utf-8")`: decodes a stream of encoded UTF-8 bytes into the native "unicode" type of the Python version.
`.encode("utf-8")`: encodes the native unicode type of the Python version to a sequence UTF-8 bytes

To summarize:
- encode("utf-8") makes something into UTF-8 encoding -> the output are UTF-8 byte sequences.
- decode("utf-8") takes UTF-8 encoded bytes and outputs a data type where you can just use "characters" ("unicode")

I can write a test case, and I hoped that the comments that I added to the function would help.

Except for converting other data types using `str(arg)` and passing None as,
it is the same as `six.ensure_str()`, which is the reference implementation of it:
https://github.com/benjaminp/six/blob/master/six.py#L916
```py
def ensure_str(s, encoding='utf-8', errors='strict'):
    """Coerce *s* to `str`.

    For Python 2:
      - `unicode` -> encoded to `str`
      - `str` -> `str`

    For Python 3:
      - `str` -> `str`
      - `bytes` -> decoded to `str`
    """
    # Optimization: Fast return for the common case.
    if type(s) is str:
        return s
    if PY2 and isinstance(s, text_type):
        return s.encode(encoding, errors)
    elif PY3 and isinstance(s, binary_type):
        return s.decode(encoding, errors)
    elif not isinstance(s, (text_type, binary_type)):
        raise TypeError("not expecting type '%s'" % type(s))
    return s
```
It does the same conditional encode/decode depending on the Python "World" it lives in.

I just did the checks a bit differently, but I actually derived it from this code.

A unit test of it could use `six.ensure_str()` as the reference which it must match.

Reply:
Unfortunately both functions seem to do the wrong thing for python2, it seems to have an 'str' that contains the individual bytes (so not unicode chars, and not UTF-8 encoded either, at least not when you'd iterate over it and print it):
```
('T\xc3\xb6r\xc3\xb6k', <type 'str'>)
('T\xc3\xb6r\xc3\xb6k', <type 'str'>)
('T\xc3\xb6r\xc3\xb6k', <type 'str'>)
('T\xc3\xb6r\xc3\xb6k', <type 'str'>)
```

Python3 seems to do the right thing:
```
Török <class 'str'>
Török <class 'str'>
Török <class 'str'>
Török <class 'str'>
```

The output is very different, but maybe I'm using the wrong print function for Python2?

Here is what I tried (basically a concatenation of the 2 functions here):
```
#!/usr/bin/env python
# -*- coding: utf-8

import sys
if sys.version_info >= (3, 0):
    text_type = str
    binary_type = bytes
    PY3 = True
else:
    text_type = unicode  # pyright:ignore[reportUndefinedVariable] # pylint: disable=unicode-builtin
    binary_type = str
    PY3 = False

PY2 = not PY3

def ensure_str(s, encoding='utf-8', errors='strict'):
    """Coerce *s* to `str`.

    For Python 2:
      - `unicode` -> encoded to `str`
      - `str` -> `str`

    For Python 3:
      - `str` -> `str`
      - `bytes` -> decoded to `str`
    """
    # Optimization: Fast return for the common case.
    if type(s) is str:
        return s
    if PY2 and isinstance(s, text_type):
        return s.encode(encoding, errors)
    elif PY3 and isinstance(s, binary_type):
        return s.decode(encoding, errors)
    elif not isinstance(s, (text_type, binary_type)):
        raise TypeError("not expecting type '%s'" % type(s))
    return s

def convert_anything_to_str(arg):
    """Converts anything into the native "str" type of the Python version, without u'str' or b'str'"""
    if arg is None or isinstance(arg, str):
        return arg  # Already str or None (checked by some callers), return it as-is:

    # If "unicode" text or "bytes", en- or decode accordingly:
    if isinstance(arg, (text_type, bytes)):
        if sys.version_info > (3, 0):
            # Python3: Decode UTF-8 bytes into the native Python3 Unicode string:
            return arg.decode("utf-8")
        else:
            # Python2: Encode the unicode text into the stream of UTF-8 bytes:
            return arg.encode("utf-8")

    # Not string-like (a number or object): Get a "str" of it using str(arg):
    return str(arg)

s1 = convert_anything_to_str(b'T\xc3\xb6r\xc3\xb6k')
s2 = convert_anything_to_str("Török")
print(s1, s1.__class__)
print(s2, s2.__class__)

s1 = ensure_str(b'T\xc3\xb6r\xc3\xb6k')
s2 = ensure_str("Török")
print(s1, s1.__class__)
print(s2, s2.__class__)
```

Reply:
Having said that I don't think XSConsole handles Unicode correctly at the moment.

Try this:
```
xe sr-param-set uuid=... name-label="Török storage"
xsconsole
```

Navigate to the 'Disk and Storage repositories' and you see this, notice the wrong encoding:
<img width="966" alt="Screenshot 2023-12-07 at 15 13 30" src="https://github.com/xapi-project/xsconsole/assets/721894/12d84a6f-72a9-4db7-a07b-144d0d13bfc8">

Reply:
and same problem if you set a VM name-label to contain unicode chars

Reply:
@edwintorok Testcase for convert_anything_to_str() added!